OpenAPI for application management

tuler opened the pull request:
Defines an openapi for application management of a (future) multi-application node.
It defines 4 endpoints: listing, adding, querying and removing.

Listing is paginated and has filtering based on status.

I used `application` instead of `dapp`. We're favoring `application` in several locations including sunodo codebase and rollups-explorer codebase, because there are several tools that generate names based on that, and the term "DApp", or "dApp" does not behave well with those tools, often generating oddities like `d-app`, or `d_app`.

gligneul commented on api/openapi.yaml:
```diff
@@ -0,0 +1,209 @@
+          enum:
+            - started
+            - starting
+            - stopping
+            - stopped
```
When do these two cases happen? Will we keep the applications as stopped once we delete them?

tuler replied:
I can remove the "stopped", but leave the "stopping".

I think there will be the case where a client will issue a DELETE, and poll for the completion of the operation (as DELETE returns 202). Then it will polling waiting for a 404 of the get. It's kind of weird.

gligneul replied:
Yes, this is weird. We can leave the stopped to avoid that.

gligneul commented on api/openapi.yaml:
```diff
@@ -0,0 +1,209 @@
+        application/json:
+          schema:
+            type: object
+            properties:
```
We should probably also add:
- epoch duration
- whether the app will run in reader mode

tuler replied:
I don't think the epoch duration whould be part of the API, for simplicity. The node would be configured (offchain) and use the same duration for all application, as this is already an offchain decision. The other params (address, block, etc) are all values that will come from the blockchain.

tuler replied:
How would you define the mode? An enum with values "reader", "validator" and "full". Or a "reader" boolean and a "validator" boolean.

gligneul replied:
We won't provide a validator mode without a reader mode in Monodo because everything will be under the same database. That said, we currently use a boolean to indicate whether the node will run in reader mode. We could be consistent here. https://github.com/cartesi/rollups-node/blob/de4b8bd4d6fa86e9c6bf3e1436cafc06825ad941/internal/config/generate/Config.toml#L24

gligneul replied:
Getting the block number might be a pain. See https://levelup.gitconnected.com/how-to-get-smart-contract-creation-block-number-7f22f8952be0

Maybe viem can come to the rescue.

tuler replied:
The API will be called upon a blockchain event, the block number won't be a problem.

gligneul reviewed (changes requested):
Looking good

tuler commented on api/openapi.yaml:
```diff
@@ -0,0 +1,209 @@
+      description: List of managed applications
+      operationId: getApplications
+      parameters:
+        - name: status
```
I left intentionally as non-required. But I don't know, I think it can be required. I'll change.

tuler commented on api/openapi.yaml:
```diff
@@ -0,0 +1,209 @@
+            pattern: "^0x([0-9a-fA-F]{40})$"
+      requestBody:
+        $ref: "#/components/requestBodies/Application"
+      responses:
```
I'll add the 409, but I think the implementation should only return 409 if the params are in conflict with the running dapp. If they are the same we could return 304.

tuler commented on api/openapi.yaml:
```diff
@@ -0,0 +1,204 @@
+openapi: 3.0.0
```
Maybe the best path is `/openapi/management.yaml`

endersonmaia commented:
I found no reference to the chain the node is running?

Is it expected to a multi-dapp node to work on multiple chains, of for a single chain?

Anyway, I think some API endpoint (or response header) should expose details for the configuration of the node, the chainID being one of these details.

```http
GET /v1/status
{ "chainID":  "123", "epochDuration": "68000", "applicationCount": 10, ...}
 ```

tuler commented:
> Is it expected to a multi-dapp node to work on multiple chains, of for a single chain?

No. It's single chain.

> Anyway, I think some API endpoint (or response header) should expose details for the configuration of the node, the chainID being one of these details.

The process running the API will receive the same env variables as a node would receive.
Aiming for simplicity, I don't think we need to expose this to an endpoint.

tuler commented:
> "applicationCount": 10

Application count is part of the API in the response of the `/applications`

gligneul commented on api/management.openapi.yaml:
```diff
@@ -0,0 +1,243 @@
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Application"
+        "400":
```
Could you reuse common responses? See https://swagger.io/docs/specification/describing-responses/#reuse

tuler replied:
Done

endersonmaia commented on api/management.openapi.yaml:
```diff
@@ -0,0 +1,243 @@
+            $ref: "#/components/schemas/Address"
+      responses:
+        "202":
+          description: Delete request accepted
```
This response could have a `Link: ` header with the status of the request that could be polled.

tuler replied:
But there would be no link to refer to. What would it be?

endersonmaia replied:
- /delete-requests/123
- /jobs/123

tuler replied:
It becomes too complicated

tuler commented on api/management.openapi.yaml:
```diff
@@ -0,0 +1,238 @@
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Application"
+        "304":
```
I think I'll remove this 304. It's not common to return 304 on a PUT request. It's usually used on a GET with `If-Modified-Since`.
I think the correct way is to just return 200 if the resource is exactly the same.

gligneul reviewed (dismissed):
Looks good!

tuler committed:
feat: openapi for application management

tuler commented:
Rebased, squashed, renamed file.

gligneul commented:
>  I'm working on the implementation, and it looks like the case will be that we will send a SIGTERM to gracefully shutdown a node, and wait for it to die with some timeout. If it succeeds we can return 204, if it times out we can return a 202. Does it make sense?

So, will it be a sync request? Do you plan to remove the `stopping` and `stopped` status?

tuler commented:
> So, will it be a sync request? Do you plan to remove the `stopping` and `stopped` status?

It will be a sync request. If times out the client would have to query the GET until it returns `stopped`, or if we remove the `stopped` until it returns 404.

gligneul commented:
> It will be a sync request. If times out the client would have to query the GET until it returns stopped, or if we remove the stopped until it returns 404.

I see. I don't have strong opinions about the stopped, you can leave it if you want.

tuler commented:
`stopped` status removed

gligneul committed:
refactor: remove unnecessary Rust code

Only the authority-claimer service will remain in the Node 2.0.

gligneul committed:
refactor: simplify CI workflows

Move the jobs from the assess-code-quality workflows to the build
workflow.

gligneul committed:
chore: remove server-manager

gligneul committed:
feat: bump contracts to 2.0-rc.2

marcelstanley committed:
fix(test): remove config for host mode

marcelstanley committed:
chore(test): run end-to-end tests via make

fmoura committed:
chore: Bump Rust to 1.78.0

Added general '-A clippy::mixed_attributes_style' to CI clippy execution to avoid linter warning on the automatically generated file 'iconsensus.rs'

GMKrieger committed:
feat(graphql): add postraphile to supervisor

gligneul committed:
fix(claimer): parse iconsensus address

fmoura committed:
feat: Add CI Custom Image

marcelstanley committed:
chore: fix typos

fmoura committed:
feat: Add --no-mining opt to Devnet

renan061 committed:
refactor: extract linewriter to package

renan061 committed:
fix: stop end-to-end testing

renan061 committed:
chore: update machine version and devnet addresses

renan061 committed:
feat: update the Dockerfile with emulator dependencies

- Adds linux.bin and rootfs.ext2 download to the emulator stage.
- Adds a emulator-devel stage that install libcmt and xgenext2fs.
- Configures the CI image to use CGO.
- Refactors the Dockerfile to improve readability.

marcelstanley committed:
feat: normalize bool config parameters

Applied suffixes ENABLED to boolean parameters to match the node
internal state and set their appropriate default values.

marcelstanley committed:
fix: adjust boolean env vars after normalization

GMKrieger committed:
fix: update postgraphile command and lock version

GMKrieger committed:
feat(data): add base repository migrations

GMKrieger committed:
feat(node): update node models

GMKrieger committed:
feat(data): add base repository functions

GMKrieger committed:
feat(data): add base repository tests

GMKrieger committed:
feat(data): add validator repository functions

GMKrieger committed:
feat(data): add validator repository tests

GMKrieger committed:
feat(data): add input reader repository functions

GMKrieger committed:
feat(data): add input reader repository tests

GMKrieger committed:
feat(node): add run migrations to node startup

GMKrieger committed:
feat(data): add postgraphile views

marcelstanley committed:
chore: fix file format

fmoura committed:
fix(tests): update port usage for testcontainers

fmoura committed:
ci: disable reaper to avoid failure on testcontainers

This issue seems related to
https://github.com/testcontainers/testcontainers-go/issues/2172

fmoura committed:
fix(tests): mine blocks while adding inputs

vfusco commented:
The management API will be a json-rpc api. It will be based on what the node cli can do, that is more than what was described here. Closing so this don't cause confusion. It's still referred on the issue #255
